src/hydro_system.hpp

```diff
@@ -583,6 +571,92 @@ void HydroSystem<problem_t>::FlattenShocks(amrex::MultiFab const &q_mf, amrex::M
 	});
 }
 
+// to ensure that physical quantities are within reasonable
+// floors and ceilings which can be set in the param file
+template <typename problem_t>
+void HydroSystem<problem_t>::EnforceLimits(amrex::Real const densityFloor, amrex::Real const pressureFloor, amrex::Real const speedCeiling,
+					   amrex::Real const tempCeiling, amrex::Real const tempFloor, amrex::MultiFab &state_mf)
+{
+
+	amrex::Real const rho_floor = densityFloor; // workaround nvcc bug
+	amrex::Real const P_floor = pressureFloor;
+	auto state = state_mf.arrays();
+
+	amrex::ParallelFor(state_mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+		int dindex = HydroSystem<problem_t>::density_index;
+		int p1index = HydroSystem<problem_t>::x1Momentum_index;
+		int p2index = HydroSystem<problem_t>::x2Momentum_index;
+		int p3index = HydroSystem<problem_t>::x3Momentum_index;
+		int eindex = HydroSystem<problem_t>::energy_index;
+		int eint_index = HydroSystem<problem_t>::internalEnergy_index;
+
+		amrex::Real const rho = state[bx](i, j, k, dindex);
+		amrex::Real const vx1 = state[bx](i, j, k, p1index) / rho;
+		amrex::Real const vx2 = state[bx](i, j, k, p2index) / rho;
+		amrex::Real const vx3 = state[bx](i, j, k, p3index) / rho;
+		amrex::Real const vsq = (vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real const Etot = state[bx](i, j, k, eindex);
+		amrex::Real const Eint = state[bx](i, j, k, eint_index);
+		amrex::Real const Temp = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Eint);
+
+		amrex::Real rho_new = rho;
+		if (rho < rho_floor) {
+			rho_new = rho_floor;
+			state[bx](i, j, k, dindex) = rho_new;
+			state[bx](i, j, k, eint_index) = Eint * rho_new / rho;
+			state[bx](i, j, k, eindex) = rho_new * vsq + state[bx](i, j, k, eint_index);
```

Comment:
Is this missing a factor of 0.5 in the kinetic energy?

Reply:
We also don't want to reset the total energy based on the auxiliary internal energy here. That should only be done inside `SyncInternalEnergy`.

Reply:
Do you mean SyncDualEnergy? So, should I call SyncDualEnergy at the end of this function?

Reply:
Update- throughout this function, I keep the Eint and Etot, Ekin separate. 
That is - `Etot_new =  Ekin_new + (Etot_old - Ekin_old)`. 
I think it is important to make sure that energy is updated as and when either density or any of the speeds are changed. This is because EnforceLimits is called several times throughout the time step and I've found that not being consistent this leads to unphysical states.

Eint is updated separately by estimating the auxiliary temperature and using the routines sitting in EOS.hpp.